src/main/java/org/opensrp/service/ImportBulkDataService.java

```diff
@@ -0,0 +1,292 @@
+		Integer rowsProcessed = 0;
+		Integer rowsInCsv = csvOrganizations.size();
+		csvBulkImportDataSummary.setNumberOfCsvRows(rowsInCsv);
+		Integer rowCount = 0;
+		List<FailedRecordSummary> failedRecordSummaries = new ArrayList<>();
+
+		Practitioner practitioner;
+		String practitionerIdentifier;
+		String practitionerName;
+		String userId;
+		String userName;
+		String organizationIdentifier;
+		String code;
+
+		String organizationIdFromCSV;
+		Long organizationId;
+		PractitionerRole practitionerRole;
+		for (Map<String, String> csvdata : csvOrganizations) {
+			try {
+				rowCount++;
+				organizationIdFromCSV = getValueFromMap(ORGANIZATION_ID_KEY, csvdata);
+				organizationId = Long.valueOf(organizationIdFromCSV);
+
+				organizationIdentifier = getOrganizationIdentifier(organizationId);
+				userName = getValueFromMap(USER_NAME_KEY, csvdata);
+				practitioner = practitionerService.getPractionerByUsername(userName);
+
+				if (validateOrganizationFields(csvdata)) {
+					practitionerIdentifier =
+							practitioner != null ? practitioner.getIdentifier() : UUID.randomUUID().toString();
+					if (practitioner == null) {
+						practitioner = new Practitioner();
+						practitioner.setIdentifier(practitionerIdentifier);
+					}
+					practitioner.setName(practitionerIdentifier);
+					practitioner.setActive(Boolean.TRUE);
+					if (csvdata.containsKey(NAME_KEY)) {
+						practitionerName = getValueFromMap(NAME_KEY, csvdata);
+					} else {
+						practitionerName = getValueFromMap(USER_NAME_KEY, csvdata);
+					}
+
+					practitioner.setName(practitionerName);
+
+					userId = getValueFromMap(USER_ID_KEY, csvdata);
+					practitioner.setUserId(userId);
+
+					userName = getValueFromMap(USER_NAME_KEY, csvdata);
+					practitioner.setUsername(userName);
+
+					practitionerRole = convertToPractitionerRole(practitionerIdentifier, csvdata);
+					if (organizationIdentifier != null) {
+						code = practitionerRole.getCode() != null ? practitionerRole.getCode().getText() : null;
+						practitionerRole.setOrganizationIdentifier(organizationIdentifier);
+						practitionerService.addOrUpdatePractitioner(practitioner);
+						practitionerRoleService
+								.assignPractitionerRole(organizationId, practitionerIdentifier, code, practitionerRole);
+						rowsProcessed++;
+					}
+				} else if (organizationIdentifier == null) {
+					failedRecordSummaries.add(getFailedRecordSummaryObject(rowCount,"Failed to get organization against the given organization Id"));
+				} else {
+					failedRecordSummaries.add(getFailedRecordSummaryObject(rowCount,"Validation failed, provided organization name mismatches with the system"));
+				}
+			}
+			catch (Exception e) {
+				logger.error(
+						"Exception occurred while persisting practitioner and pratitioner role data : " + e.getMessage());
+				failedRecordSummaries.add(getFailedRecordSummaryObject(rowCount, "Exception occurred due to : " + e.getMessage()));
```

Comment:
is it possible to know the type of error here? can we use know errors and return constant strings. This may cause information leak

Reply:
Could you please clarify what does it mean? The reason for adding failedRecordSummaries in the catch block is due to the check on all the failed imports as we are displaying all the summary details in the CSV file to the user in response. It could be in the catch block due to any exception thrown while persisting the data.

Reply:
I am recommending we return a custom error message in `e.getMessage())` in `failedRecordSummaries.add(getFailedRecordSummaryObject(rowCount, "Exception occurred due to : " + e.getMessage()));`

Reply:
To handle this issue, should I return an error message as an "Unknown error occurred" instead of displaying the exact exception thrown by the service layer to prevent information leak? @githengi

Reply:
Yes for exceptions other than expected. You can have catch block for IllegalArg,validation exceptions, this can return the exception message, the another block that catches all exceptions which returns "Unknown error occurred".

Reply:
Thanks for clarifying. All validation errors are handled outside catch block, will update this catch block error message to "Unknown error occurred".